### PR TITLE
Support shared headroom pool on top of dynamic buffer calculation 

### PR DESCRIPTION
**Pull request by stephenxs**
### Description of PR

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

Signed-off-by: Stephen Sun <stephens@nvidia.com>

#### What is the motivation for this PR?
Add test cases for shared headroom pool on top of the dynamic buffer calculation.

#### How did you do it?
1. Update the existing test cases for the dynamic buffer calculation in the following aspects:
- When checking the pool size, take shared headroom pool related parameters into account
- Replace "headroom" with "size" as the it stands for size of the PG.
2. Add test case for shared headroom pool configuration test.

#### How did you verify/test it?
Run the regression test.

### CODE REVIEW

**Comment by stephenxs** [comment]
retest this, please

**Comment by liat-grozovik** [comment]
@stephenxs please review LGTM errors and lets clean new ones added.

**Comment by stephenxs** [comment]
> @stephenxs please review LGTM errors and lets clean new ones added.

There is no new LGTM alert introduced by the latest commit. What it means is that an **existing** alert has been **fixed**.

**Comment by stephenxs** [comment]
retest this please

**Comment by keboliu** [comment]
retest vsimage please

**Comment by stephenxs** [comment]
retest this, please

**Comment by stephenxs** [comment]
retest vsimage please

**Comment by keboliu** [comment]
retest vsimage please

**Comment by stephenxs** [comment]
retest this, please

**Comment by liat-grozovik** [comment]
reste vsimage please

**Comment by liat-grozovik** [comment]
retest vsimage please

**Comment by stephenxs** [comment]
retest vsimage please

**Comment by stephenxs** [comment]
retest this please

**Comment by liat-grozovik** [comment]
@neethajohn  could you please review and provide your feedback?

**Comment by liat-grozovik** [comment]
retest vsimage please

File: tests/qos/test_buffer.py [hunk 1]
```diff
@@ -141,61 +170,135 @@ def check_pool_size(duthost, ingress_lossless_pool_oid, **kwargs):
         kwargs: The parameters based on which the expected pool size is calculated.
                 They are represeted in form of kwargs because different vendor can require different parameters
```
Comment (neethajohn) [inline tests/qos/test_buffer.py:171]: 'represented'

File: tests/qos/test_buffer.py [hunk 2]
```diff
@@ -141,61 +170,135 @@ def check_pool_size(duthost, ingress_lossless_pool_oid, **kwargs):
...
 
-        if "old_pg_number" in kwargs:
-            old_pg_number = int(kwargs["old_pg_number"])
+    if duthost.facts['asic_type'] == 'mellanox':
+        if kwargs.get("old_ratio") and kwargs.get("new_ratio"):
+            curr_pool_size = int(kwargs["pool_size"])
+            curr_shp_size = int(kwargs["shp_size"])
+            old_ratio = int(kwargs.get("old_ratio"))
+            new_ratio = int(kwargs.get("new_ratio"))
+            original_memory = curr_pool_size * DEFAULT_INGRESS_POOL_NUMBER + curr_shp_size
```
Comment (neethajohn) [inline tests/qos/test_buffer.py:194]: Will this logic work when there are multiple ingress pools and they do not have the same size?

File: tests/qos/test_buffer.py [hunk 3]
```diff
@@ -141,61 +170,135 @@ def check_pool_size(duthost, ingress_lossless_pool_oid, **kwargs):
...
+            if new_ratio == 0:
+                expected_shp_size = 0
```
Comment (neethajohn) [inline tests/qos/test_buffer.py:196]: why is this 0? even if ratio is 0, if shp_size is set, then shared pool is enabled and we should expect that size?

File: tests/qos/test_buffer.py (hunk 2 above)
Comment (stephenxs) [reply tests/qos/test_buffer.py:194]: No. But this is implemented on a per vendor basis and that is not the case for Mellanox.

File: tests/qos/test_buffer.py [hunk 4]
```diff
@@ -46,6 +51,27 @@ def detect_ingress_pool_number(duthost):
...
+    global DEFAULT_SHARED_HEADROOM_POOL_ENABLED
+    global DEFAULT_SHARED_HEADROOM_POOL_SIZE
+    global DEFAULT_OVER_SUBSCRIBE_RATIO
+
+    over_subscribe_ratio = duthost.shell('redis-cli -n 4 hget "DEFAULT_LOSSLESS_BUFFER_PARAMETER|AZURE" over_subscribe_ratio')['stdout']
+    if over_subscribe_ratio and  over_subscribe_ratio != '0':
+        DEFAULT_SHARED_HEADROOM_POOL_ENABLED = True
+        DEFAULT_OVER_SUBSCRIBE_RATIO = int(over_subscribe_ratio)
+
+    shared_headroom_pool_size = duthost.shell('redis-cli -n 4 hget "BUFFER_POOL|ingress_lossless_pool" size')['stdout']
```
[the remark this answers is not in the document]
Comment (stephenxs) [reply tests/qos/test_buffer.py:69]: It’s called at the beginning of the test for the purpose of detecting SHP mode. We don’t enable it by setting size in our test beds at the beginning of the test. This is why we don’t observe this typo.

File: tests/qos/test_buffer.py (hunk 3 above)
Comment (stephenxs) [reply tests/qos/test_buffer.py:196]: This is to verify the SHP size when the following factors are changed:
- over subscribe ratio
- Lossless PGs 

the case SHP is configured by setting size is handled in the next elif branch (config_shp_size)

**Review by neethajohn — APPROVED** [review]

**Pull request merged**